### ISSUE TITLE
question about [The one-line change]

### ISSUE THREAD

**Issue by dongdk** [csy2077/data-forcing-distillation#2]
hi @csy2077 
in paper pseudo code is
`teacher_in = data.detach() if torch.rand() < p else gen.detach()`， but [in code](https://github.com/csy2077/data-forcing-distillation/blob/7281906026276dc46e23eb92b72238ebc050b463/fastgen/methods/distribution_matching/dmd2.py#L250), it uses detach op to implement. why to do this?

**Comment by csy2077** [comment]
Good question — the two are equivalent, the code just expresses it with a straight-through trick instead of a literal input switch.

The line

```python
gen_data_teacher_score = gen_data + (real_sample - gen_data).detach()
```

is the stop-gradient equation `x + (target - x).detach()`. In the forward pass the two `gen_data` terms cancel, so `gen_data_teacher_score` numerically **equals `real_sample`** — the teacher (and the fake score, when `forward_kl=True`) gets evaluated at the real data point. That's exactly the `teacher_in = data` branch from the pseudocode. The `else` branch (probability `1 − p`) is just the default `gen_data_teacher_score = gen_data`.

Why write it this way rather than a plain `= real_sample`:

- It keeps the variable in the autograd graph with the same dtype/device/shape as `gen_data`, so it's a clean drop-in — this matters most in the partial-replacement case `gen_data_teacher_score[gen_idx] = gen_data_teacher_score[gen_idx] + (real_sample - gen_data[gen_idx]).detach()`, where only one batch element is swapped and an in-place assignment of a real tensor into a graph tensor would risk breaking autograd. That partial-replacement branch was actually the early trial version of the DFD method, and we just kept the same code structure for the full version, so both paths share this straight-through form.
- It's the general/robust form: if the teacher output were ever left attached, gradient would still route to `gen_data` with identity Jacobian.

On the `.detach()` question specifically — why it's equivalent to the pseudocode's `data.detach()`: data-forcing only needs the real sample's **value**, not a gradient through it. Trace the consumers of `gen_data_teacher_score`: it feeds `self.teacher(...)` (returned as `teacher_x0_teacher_score.detach()`) and `self.fake_score(...)` (under `torch.no_grad()`). Both terminate in a detach / no_grad, and the VSD loss is differentiated w.r.t. the *original* `gen_data`, not `gen_data_teacher_score`:

```python
vsd_loss = variational_score_distillation_loss(gen_data, teacher_x0, fake_score_x0, forward_kl_ensemble)
```

Inside that loss, everything up to `pseudo_target = gen_data - vsd_grad` is under `no_grad`, so the actual gradient delivered to `gen_data` is just the constant `vsd_grad = (fake_score_x0 - teacher_x0) * w`. Data-forcing enters only by changing the *value* of `teacher_x0` (since it's now evaluated at the real sample), which shifts that constant. No gradient ever needs to flow through the teacher's input — which is why both the pseudocode's `.detach()` and the code's straight-through `.detach()` give identical updates.

So: same behavior, and the straight-through form is just a safer, drop-in way to route the real value into the teacher evaluation while leaving the optimized quantity (`gen_data` in the MSE) untouched.

I hope this could answer your question :)
